### PR TITLE
HMIS DQ report

### PR DESCRIPTION
**Pull request by eanders**

### CODE REVIEW

File: app/models/filters/filter_base.rb [hunk 1]
```diff
@@ -408,6 +410,44 @@ def anded_effective_project_ids
       ids.reject(&:empty?).reduce(&:&)
     end
 
+    # Apply all known scopes
+    # NOTE: by default we use coc_codes, if you need to filter by the coc_code singular, make note
+    def apply(scope, all_project_types: nil, use_coc_codes: true, use_coc_code: false)
```
Comment (awisema) [inline app/models/filters/filter_base.rb]: Does `use_coc_codes: true, use_coc_code: true` ever make sense? If not, maybe something like `filter_by_coc: true, multiple_coc_codes: true`?

File: app/views/filters/show.haml [hunk 2]
```diff
@@ -2,4 +2,5 @@
   Selected Universe
 
 .report-parameters-all
+  = debug filter_params
```
Comment (awisema) [inline app/views/filters/show.haml]:
```suggestion
```

File: drivers/hmis_data_quality_tool/README.md [hunk 3]
```diff
@@ -0,0 +1,3 @@
+## HmisDataQualityTool README
+
+This README file should be used to explain the functionality of the driver.
```
Comment (awisema) [inline drivers/hmis_data_quality_tool/README.md]:
```suggestion
The HMIS Data Quality tool is a screening tool to identify data quality issues that will affect the LSA or similar analyses. It is based on the [HMIS CSV Data Quality (DQ) Tool](https://bitbucket.org/abtassociates/hmis-csv-dq-tool/src/main/)
```

File: drivers/hmis_data_quality_tool/app/controllers/hmis_data_quality_tool/warehouse_reports/reports_controller.rb [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+      HmisDataQualityTool::Report
+    end
+
+    private def set_filter
+      @filter = filter_class.new(user_id: current_user.id, enforce_one_year_range: false, require_service_during_range: false)
+      @filter.update(filter_params[:filters]) if filter_params[:filters].present?
+    end
+
+    def filter_params
+      site_coc_codes = GrdaWarehouse::Config.get(:site_coc_codes).presence&.split(/,\s*/)
```
Comment (awisema) [inline drivers/hmis_data_quality_tool/app/controllers/hmis_data_quality_tool/warehouse_reports/reports_controller.rb]: Seems like we should standardize this somehow -- even if only add `def get_as_array(...)` to `Config`?

File: drivers/hmis_data_quality_tool/app/models/hmis_data_quality_tool/client.rb [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+      report_item.am_ind_ak_native = client.AmIndAKNative
+      report_item.asian = client.Asian
+      report_item.black_af_american = client.BlackAfAmerican
+      report_item.native_hi_pacific = client.NativeHIPacific
+      report_item.white = client.White
+      report_item.race_none = client.RaceNone
+      # we need these for calculations, but don't want to store them permanently
+      report_item.enrollments = client.source_enrollments
+      report_item.overlapping_entry_exit = overlapping_entry_exit(enrollments: report_item.enrollments, report: report)
+      # FIXME
```
Comment (awisema) [inline drivers/hmis_data_quality_tool/app/models/hmis_data_quality_tool/client.rb]: What needs fixing here?

File: drivers/hmis_data_quality_tool/app/models/hmis_data_quality_tool/enrollment.rb [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+      report_item.household_max_age = hh.map { |en| en[:age] }.compact.max
+      report_item.head_of_household_count = hh.select { |en| en[:relationship_to_hoh] == 1 }.count
+
+      max_date = [report.filter.end, Date.current].min
+      en_services = enrollment.services&.select { |s| s.DateProvided <= max_date }
+      en_cls = enrollment.current_living_situations&.select { |s| s.InformationDate <= max_date }
+
+      lot = if project_tracking_method == 3
+        # count services <= min of report end and current date
+        en_services&.count || 0
```
Comment (awisema) [inline drivers/hmis_data_quality_tool/app/models/hmis_data_quality_tool/enrollment.rb]: Should this be limited to bed nights?

File: drivers/hmis_data_quality_tool/app/models/hmis_data_quality_tool/enrollment.rb [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+            return false if item.living_situation.blank?
+
+            ! item.living_situation.in?(HUD.valid_prior_living_situations)
+          },
+        },
+        exit_date_issues: {
+          title: 'Exit before Entry',
+          description: 'Enrollment exit date must occur after entry date',
+          limiter: ->(item) {
+            item.exit_date.present? && item.exit_date < item.entry_date
```
Comment (awisema) [inline drivers/hmis_data_quality_tool/app/models/hmis_data_quality_tool/enrollment.rb:178]: Is it also an error if entry and exit are the same day?
```suggestion
            item.exit_date.present? && item.exit_date <= item.entry_date
```

File: drivers/hmis_data_quality_tool/app/models/hmis_data_quality_tool/enrollment.rb [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+            return false if item.exit_date.blank?
+
+            ! HUD.valid_destinations.key?(item.destination)
+          },
+        },
+        unaccompanied_youth_issues: {
+          title: 'Unaccompanied Youth < 12 Years Old',
+          description: 'Youth under 12 are generally expected to be accompanied.  The presence of an unaccompanied youth under 12 may indicate an issue with household data collection',
+          limiter: ->(item) {
+            item.household_max_age.present? && item.household_max_age <= 12
```
Comment (awisema) [inline drivers/hmis_data_quality_tool/app/models/hmis_data_quality_tool/enrollment.rb]: I think this is HUDs fuzziness in the description, but this implements under the age of 13, right?

File: drivers/hmis_data_quality_tool/app/models/hmis_data_quality_tool/enrollment.rb [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+        days_since_last_service_so_365_issues: {
+          title: 'Possible Missed Exit - SO, Time in Enrollment 365 Days or More',
+          description: 'There is an expectation that clients will not stay indefinitely in street outreach, these clients have been in street outreach with no current living situation collected for more than 365 days',
+          limiter: ->(item) {
+            item.days_since_last_service > 365 && GrdaWarehouse::Hud::Project::RESIDENTIAL_PROJECT_TYPES[:so].include?(item.project_type)
+          },
+        },
+        days_in_ph_prior_to_move_in_90_issues: {
+          title: 'Possible Missed Move In Date - PH, Time in Enrollment 90 Days or More',
+          description: 'There is an expectation that clients in PH will eventually move into houseing, these clients have been in PH without a move-in date or with an invalid move-in date more than 90 days',
```
Comment (awisema) [inline drivers/hmis_data_quality_tool/app/models/hmis_data_quality_tool/enrollment.rb]: I expect this is from the tool, but:
```suggestion
          description: 'There is an expectation that clients in PH will eventually move into housing, these clients have been in PH without a move-in date more than 90 days,  or have an invalid move-in date ',
```
Same edit would apply for 180 and 365 days.

**Review by awisema — CHANGES_REQUESTED** [review]
Logic all makes sense to me -- some minor tweaks

File: drivers/hmis_data_quality_tool/app/controllers/hmis_data_quality_tool/warehouse_reports/reports_controller.rb (hunk 4 above)
Comment (eanders) [reply drivers/hmis_data_quality_tool/app/controllers/hmis_data_quality_tool/warehouse_reports/reports_controller.rb]: good idea.

File: drivers/hmis_data_quality_tool/README.md (hunk 3 above)
Comment (eanders) [reply drivers/hmis_data_quality_tool/README.md]: Thanks

File: app/views/filters/show.haml (hunk 2 above)
Comment (eanders) [reply app/views/filters/show.haml]: whoops!

File: app/models/filters/filter_base.rb (hunk 1 above)
Comment (eanders) [reply app/models/filters/filter_base.rb]: I don't think so.  For now, adding a single argument to the method.

File: drivers/hmis_data_quality_tool/app/models/hmis_data_quality_tool/enrollment.rb (hunk 6 above)
Comment (eanders) [reply drivers/hmis_data_quality_tool/app/models/hmis_data_quality_tool/enrollment.rb]: let's do that.

File: drivers/hmis_data_quality_tool/app/models/hmis_data_quality_tool/enrollment.rb (hunk 7 above)
Comment (eanders) [reply drivers/hmis_data_quality_tool/app/models/hmis_data_quality_tool/enrollment.rb:178]: I'm leaving this as <=.  Only residential projects have this limitation, not sure if we should enforce that here or not.

File: drivers/hmis_data_quality_tool/app/models/hmis_data_quality_tool/enrollment.rb (hunk 8 above)
Comment (eanders) [reply drivers/hmis_data_quality_tool/app/models/hmis_data_quality_tool/enrollment.rb]: correct, should be < 12

File: drivers/hmis_data_quality_tool/app/models/hmis_data_quality_tool/enrollment.rb (hunk 9 above)
Comment (eanders) [reply drivers/hmis_data_quality_tool/app/models/hmis_data_quality_tool/enrollment.rb]: Good catches.  Thanks!

**Pull request merged**